File: aznavrail-react/src/guidance/AzCalloutPlacement.ts
import type { AzShapeBounds } from './AzStatus';

/**
 * Pure geometry for placing a guidance callout near its target but never on top of it, off the known
 * obstacles, and inside the safe area. Window-space px. Mirrors the Kotlin `AzCalloutPlacement`.
 */

/** Area of the intersection of two rects (0 when disjoint). */
export function overlapArea(a: AzShapeBounds, b: AzShapeBounds): number {
  const w = Math.max(0, Math.min(a.left + a.width, b.left + b.width) - Math.max(a.left, b.left));
  const h = Math.max(0, Math.min(a.top + a.height, b.top + b.height) - Math.max(a.top, b.top));
  return w * h;
}

function clampToSafe(x: number, y: number, w: number, h: number, safe: AzShapeBounds): AzShapeBounds {
  const maxLeft = Math.max(safe.left, safe.left + safe.width - w);
  const maxTop = Math.max(safe.top, safe.top + safe.height - h);
  const left = Math.min(Math.max(x, safe.left), maxLeft);
  const top = Math.min(Math.max(y, safe.top), maxTop);
  return { left, top, width: w, height: h };
}

function penalty(rect: AzShapeBounds, obstacles: AzShapeBounds[], target: AzShapeBounds | null): number {
  let p = 0;
  if (target) p += overlapArea(rect, target) * 2;
  for (const o of obstacles) p += overlapArea(rect, o);
  return p;
}

/**
 * Choose the best window-space rect for a `size` callout: the four sides of `target` (with `gap`), or a
 * grid scan of `safe` when there is no target. Lowest obstacle-overlap wins (ties prefer the earlier
 * candidate — below first), then it is clamped fully into `safe`.
 */
export function placeCallout(
  target: AzShapeBounds | null,
  size: { width: number; height: number },
  obstacles: AzShapeBounds[],
  safe: AzShapeBounds,
  gap = 8,
): AzShapeBounds {
  const w = size.width;
  const h = size.height;
  const candidates: AzShapeBounds[] = [];
  if (target) {
    candidates.push(clampToSafe(target.left, target.top + target.height + gap, w, h, safe)); // below
    candidates.push(clampToSafe(target.left, target.top - h - gap, w, h, safe)); // above
    candidates.push(clampToSafe(target.left + target.width + gap, target.top, w, h, safe)); // end
    candidates.push(clampToSafe(target.left - w - gap, target.top, w, h, safe)); // start
  } else {
    const stepX = Math.max(1, w * 0.5);
    const stepY = Math.max(1, h * 0.5);
    const maxY = Math.max(safe.top, safe.top + safe.height - h);
    const maxX = Math.max(safe.left, safe.left + safe.width - w);
    for (let y = safe.top; y <= maxY; y += stepY) {
      for (let x = safe.left; x <= maxX; x += stepX) {
        candidates.push(clampToSafe(x, y, w, h, safe));
      }
    }
    if (candidates.length === 0) candidates.push(clampToSafe(safe.left, safe.top, w, h, safe));
  }
  let best = candidates[0];
  let bestP = penalty(best, obstacles, target);
  for (let i = 1; i < candidates.length; i++) {
    const p = penalty(candidates[i], obstacles, target);
    if (p < bestP) {
      best = candidates[i];
      bestP = p;
    }
  }
  return best;
}

File: aznavrail-cmp/src/commonMain/kotlin/com/hereliesaz/aznavrail/tutorial/AzCalloutPlacement.kt
package com.hereliesaz.aznavrail.tutorial

import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size

/**
 * Pure geometry for placing a guidance callout so it sits **near its target but never on top of it**,
 * avoids the known [obstacles] (other rail items / targets / already-placed callouts), and stays fully
 * inside the [safe] area (so text is never clipped and nothing lands in the system-bar zone). All values
 * are window-space px. Kept pure so it is unit-testable without a renderer; mirrored in the React port.
 */

/** Area of the intersection of two rects (0 when disjoint). */
internal fun overlapArea(a: Rect, b: Rect): Float {
    val w = (minOf(a.right, b.right) - maxOf(a.left, b.left)).coerceAtLeast(0f)
    val h = (minOf(a.bottom, b.bottom) - maxOf(a.top, b.top)).coerceAtLeast(0f)
    return w * h
}

/** Shift a [width]×[height] rect anchored at ([x],[y]) so it lies fully within [safe] (best effort). */
private fun clampToSafe(x: Float, y: Float, width: Float, height: Float, safe: Rect): Rect {
    val left = x.coerceIn(safe.left, (safe.right - width).coerceAtLeast(safe.left))
    val top = y.coerceIn(safe.top, (safe.bottom - height).coerceAtLeast(safe.top))
    return Rect(left, top, left + width, top + height)
}

/** Total overlap of [rect] against every obstacle plus the [target] it must not cover. */
private fun penalty(rect: Rect, obstacles: List<Rect>, target: Rect?): Float {
    var p = 0f
    if (target != null) p += overlapArea(rect, target) * 2f // weigh covering the target itself heaviest
    obstacles.forEach { p += overlapArea(rect, it) }
    return p
}

/**
 * Choose the best window-space rect for a [size] callout. When [target] is non-null the four sides
 * (below, above, end, start, with [gap]) are tried; otherwise a coarse grid of [safe] is scanned so
 * untargeted callouts spread out instead of stacking. The lowest-overlap candidate wins (ties resolve in
 * the listed order — below is preferred), then it is clamped into [safe].
 */
internal fun placeCallout(
    target: Rect?,
    size: Size,
    obstacles: List<Rect>,
    safe: Rect,
    gap: Float = 8f,
): Rect {
    val w = size.width
    val h = size.height
    val candidates = ArrayList<Rect>()
    if (target != null) {
        // Below, above, end, start — anchored to the target, then clamped into the safe area.
        candidates += clampToSafe(target.left, target.bottom + gap, w, h, safe)
        candidates += clampToSafe(target.left, target.top - h - gap, w, h, safe)
        candidates += clampToSafe(target.right + gap, target.top, w, h, safe)
        candidates += clampToSafe(target.left - w - gap, target.top, w, h, safe)
    } else {
        // No anchor: scan a grid of the safe area (so multiple untargeted callouts don't collide).
        val stepX = (w * 0.5f).coerceAtLeast(1f)
        val stepY = (h * 0.5f).coerceAtLeast(1f)
        var y = safe.top
        while (y <= (safe.bottom - h).coerceAtLeast(safe.top)) {
            var x = safe.left
            while (x <= (safe.right - w).coerceAtLeast(safe.left)) {
                candidates += clampToSafe(x, y, w, h, safe)
                x += stepX
            }
            y += stepY
        }
        if (candidates.isEmpty()) candidates += clampToSafe(safe.left, safe.top, w, h, safe)
    }
    // Pick the minimum-penalty candidate; the first listed wins ties (stable preference order).
    var best = candidates.first()
    var bestPenalty = penalty(best, obstacles, target)
    for (i in 1 until candidates.size) {
        val p = penalty(candidates[i], obstacles, target)
        if (p < bestPenalty) {
            best = candidates[i]
            bestPenalty = p
        }
    }
    return best
}
